engine-nextgen/src/main/kotlin/graphql/nadel/enginekt/transform/hydration/batch/NadelBatchHydrationByObjectId.kt

```diff
@@ -41,26 +75,46 @@ internal object NadelBatchHydrationByObjectId {
                     else -> error("Hydration actor result must be an object")
                 }
             }
-            .associateBy {
-                // We don't want to show this in the overall result, so remove it here as we use it
-                it.asMutable().remove(state.aliasHelper.getResultKey(matchStrategy.objectId))
+            .associateBy { resultNode ->
+                objectIds
+                    .map { objectId ->
+                        // We don't want to show this in the overall result, so remove it here as we use it
+                        resultNode.asMutable().remove(state.aliasHelper.getResultKey(objectId.resultId))
+                    }
             }
 
-        val resultKeysToObjectIdOnHydrationParentNode = state.aliasHelper.getQueryPath(
-            getPathToObjectIdentifierOnHydrationParentNode(instruction),
-        )
+        val pathsToSourceIds = objectIds
+            .map { objectId ->
+                state.aliasHelper.getQueryPath(objectId.sourceId)
+            }
 
-        return parentNodes.map { parentNode ->
-            val parentNodeIdentifierNodes = JsonNodeExtractor.getNodesAt(
-                rootNode = parentNode,
-                queryPath = resultKeysToObjectIdOnHydrationParentNode,
-            )
+        return parentNodes.map { sourceNode ->
+            // [
+            //   [page-1, page-2, page-1], // page id
+            //   [draft,  posted, posted] // union of [draft posted]
+            // ]
+            val sourceIdNodes = pathsToSourceIds
+                .map { path ->
+                    JsonNodeExtractor.getNodesAt(sourceNode, path)
+                        .asSequence()
+                        .map(JsonNode::value)
+                        .flatten(recursively = true)
+                        .toList()
+                }
+
+            val sourceIds = (0 until sourceIdNodes.first().size)
+                .map { keyIndex ->
+                    sourceIdNodes
+                        .map {
+                            it[keyIndex]
+                        }
+                }
```

Comment:
This is slightly confusing code that I will try to revisit, but basically the issue is that one source node can have multiple IDs for the batch call e.g.

```graphql
type ComplexIssueId {
  id: ID
  siteId: ID
}
type Issue {
  issueIds: [ComplexIssueId]
  children: [Issue] @hydrated(
    arguments: [
      {name: "ids", value: "$source.issueIds"}
    ]
    objectIdentifiedBy: [
      {sourceId: "issueIds.id" resultId: "id"},
      {sourceId: "issueIds.siteId" resultId: "site"},
    ]
  )
}
```

so when we extract `issueIds.id` AND `issueIds.siteId` we get two arrays like

```
[
  [issue-1, issue-2, issue-1], // id
  [hello,   hello,   jdog] // site id
]
```

Here we reconstruct the keys required for the `resultNodesByObjectId` map as e.g.

```
[
  [issue-1, hello]
  [issue-2, hello]
  [issue-1, jdog]
]
```